src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -3549,32 +3486,91 @@ HRESULT ThreadSuspend::SuspendRuntime(ThreadSuspend::SUSPEND_REASON reason)
     // Make a pass through all threads.  We do a couple of things here:
     // 1) we count the number of threads that are observed to be in cooperative mode.
     // 2) for threads currently running managed code, we try to redirect/jihack them.
-    //
-    // Later we will make more passes where we do roughly the same thing.  We should combine the two loops.
-    //
 
-    while ((thread = ThreadStore::GetThreadList(thread)) != NULL)
+    // counts of cooperative threads
+    int previousCount = 0;
+    int countThreads = 0;
+
+    // we will iterate over threads and check which are left in coop mode
+    // while checking, we also will try suspending and hijacking
+    // unless we have just done that, then we can observeOnly and see if situation improves
+    // we do not on uniprocessor though (spin-checking is pointless on uniprocessor)
+    bool observeOnly = false;
+
+    _ASSERTE(!pCurThread->HasThreadState(Thread::TS_GCSuspendFlags));
+#ifdef _DEBUG
+    DWORD dbgStartTimeout = GetTickCount();
+#endif
+
+    for (;;)
     {
-        if (thread->HasThreadState(Thread::TS_GCSuspendPending))
+        for (Thread *thread = ThreadStore::GetThreadList(NULL); thread != NULL; thread = ThreadStore::GetThreadList(thread))
```

Comment:
Reverted back this and 2 other places to be:
```c
    Thread* thread = NULL;
    while ((thread = ThreadStore::GetThreadList(thread)) != NULL)

```

---
In reply to: [503476859](https://github.com/dotnet/runtime/pull/34666#discussion_r503476859) [](ancestors = 503476859)

Reply:
`for` syntax could be a bit more robust for loops like this, since it scopes the iteration variable (`thread` here) to the loop body. It makes it a bit easier to track who changes it and who uses it. It is a mild preference though.

---
In reply to: [505819489](https://github.com/dotnet/runtime/pull/34666#discussion_r505819489) [](ancestors = 505819489,503476859)

Reply:
made it a while loop